### PR TITLE
[READY] Explosive Tesla, Take 2

### PR DESCRIPTION
**Pull request by PKPenguin321**
Video (a bit outdated): https://www.youtube.com/watch?v=f_CrgMbk4t4
Remake of https://github.com/tgstation/tgstation/pull/17501

**Todo:**
- [x] Make tesla itself bomb immune since it will delete itself in some cases
- [x] Make an upgrade for cameras and ~~airlocks~~ light tubes that tesla-proofs them to prevent tesla from breaking its own containment
- [x] Do a testmerge of this to get a feel for how laggy it is and adjust explosion values and probabilities accordingly

**Maps Tested On:**
- [x] Boxstation: Containment area can successfully hold a tesla ball with the equipment given at roundstart. Lights occasionally blow out, but it's inconsequential.
- [x] Metastation: Also works.
- [x] Birdboat/ministation: Doesn't even have a singulo containment area so no issue there
- [x] Dreamstation: :white_check_mark: 
- [x] EfficiencyStation :white_check_mark: Lights occasionally blow out, but it's inconsequential.

**Changes made:**
- tesla_act for machinery in general now often calls minibomb-sized explosions (which lack smoke and admin logging to reduce lag and spam, respectively)
- explosion proc now has an argument that lets you determine if you want your explosion to make smoke or not. Defaults to 1
- The effect_system itself for explosions is split into two paths, one that makes smoke and one that doesn't (should be inconsequential)
- Makes light tubes and their variants spew fire on tesla_act instead of a full explosion (to keep tesla from breaking itself out of containment)
- Makes cameras with EMP immunity upgrade also tesla immune
- Makes the tesla itself bomb immune
- Refactored tesla grilles. They no longer shoot a real tesla beam, just one that looks exactly like a tesla beam and will ONLY target carbon mobs (meaning no explosions). As a consequence, they will only ever target 1 person at a time, meaning you can't kill 40 people with one zap anymore.
- Made the gravity generator get properly destroyed by the tesla so that it doesn't just repeatedly explode over and over.
- Made the nuke do the same as the gravity generator. As a consequence, nukes can now be "disarmed" by the tesla (in the same way BoH bombing a nuke disarms it).

**Other notes:**
- I've been able to make the tesla successfully without it breaking loose on it's own. ~~on Boxstation, but other maps are yet to be tested.~~ It's been fully tested and proper engine setup should never lead to a tesloose.
- Airlocks do not need to be bomb-proofed ~~as far as I can tell~~ thanks to grounding rods being right next to them. ~~(at least on Boxstation)~~

The goal here is to make the tesla about as destructive as the singularity is, instead of EMPing and occasionally deleting some machines.

:cl: PKPenguin321
add: The tesla is now much more dangerous, and can cause electronics to explode violently.
/:cl:

### CODE REVIEW

**Comment by WJohn** [comment]
Holy shit. I've been asking for this for a while. :+1: x :100:

**Comment by Core0verload** [comment]
:-1: Tesla is already an unkillable round ender. Make it decay if you really want this.

**Comment by PKPenguin321** [comment]
If it's gonna be an unkillable round ender, might as well make it go all-out instead of just EMPing things, right? More fun that way

**Comment by Shadowlight213** [comment]
@PKPenguin321 It deletes shit by calling ex_act on it as well.

**Comment by PKPenguin321** [comment]
Yep! The singularity also deletes shit, but it just does it directly, so I guess you could say this is a step down.

**Comment by PKPenguin321** [comment]
Alright then, all that's left is to testmerge and tweak values. I'm starting with explosions being the most likely thing and them being minibomb-strong, which is pretty much the maximum I'll go with this. If it's laggy (which is likely) I will likely do one or more of the following:
- Make explosions not the most likely option
- Make multiple probabilities with weak explosions being the most common and strong being the least

**Comment by ShadowDeath6** [comment]
don't minibombs instant crit if you're point blanked with them?

now we get double instacrit tesla action, with 100% less counter memes

**Comment by PKPenguin321** [comment]
The singularity insta*gibs,* so you're still better off.

**Comment by TechnoAlchemisto** [comment]
This is a good thing, a loosed engine is supposed to be ridiculously chaotic and dangerous. Lethality memes.

**Comment by Cheridan** [comment]
Singularity instagibs if you manage to fall into it, which is surprisingly avoidable.
Tesla just pops onto your screen and instakills you and everyone you love.

**Comment by PKPenguin321** [comment]
>The difference between gibbing and killing in this situation is pretty meaningless since if the tesla is loose, the shuttle is coming. You're probably not getting cloned before the round ends.

Yep, and the difference between machines getting emagged and blown up is pretty meaningless if the shuttle is coming, so why not have the more exciting option?

**Comment by kevinz000** [comment]
10/10

**Comment by Ninja137** [comment]
It's beautiful. Make it lose charge slower, though.

**Comment by Anon-svn** [comment]
The perfect concept, actually makes the tesla coil fun to watch

**Comment by PKPenguin321** [comment]
>Do you rike it
Too strong: 1
It's perfect: 4
MAKE IT STRONGER: 17
Did not vote: 42

Testmerge observations:
- No lag!!!!!! It's perfect! :smile: 
- Make mechs asplode
- Maybe make nukes and grav gen get deleted instead of exploding over and over
- *Maybe* remove hardsuit immunity from tesla? (Will save for another PR in any case)
- Players in deadchat fucking loved it, a few said to lower the range but overall feedback was very high in deadchat
- You could hear the tesla before it got to you and started bombing, so that's good

**Comment by Shadowlight213** [comment]
One of the grounding rods got moved from its position and the tesla hit an engineer, gibbed them,  and blew a huge hole in engineering.

**Comment by TechnoAlchemisto** [comment]
I wouldn't remove the tesla immunity considering everything around you explodes.

**Comment by Ninja137** [comment]
No. What happened was this:

-Set up engine
-[REDACTED] happens
-Panic, forget what I'm doing
-Continue setting up engine
-Accidentally put in the Singulo
-Turn on after hacking PA
-Singulo moves rods
-Ask admins to replace engine with tesla
-No grounding rods
-I'm standing adjacent to the eastern collector bank
-The very first shock explodes the entire side of the room and I'm immediately disintegrated, and practically jump out of my chair

Round Highlights of Tesla:

-Maxwell Gaylord's enfeebled right leg of elf slaying -11 has been violently dismembered!
-Tesla ripping through the entirety of Engi, Atmos, Med, Sci, Viro, and Escape in under two minutes after escaping

**Comment by PKPenguin321** [comment]
I should also say that before this gets merged it needs to be tested on other maps. I know box can contain it, but I don't know about meta or any of the other maps. Map edits are likely needed (which sucks because I have no idea how to do them), it might be good to also put reinforced plating over the wires that lead to emitters, I don't know.

**Comment by Ninja137** [comment]
No man, it has to be able to escape. It can't natively escape, so it shouldn't be too massive an amount of shit to make it go kill everything.

**Comment by Iamgoofball** [comment]
@Ninja137 The singulo also can't escape by itself it requires malpractice or sabotage

I really don't like this. It completely wrecks the design of the tesla.

File: code/game/machinery/machinery.dm [hunk 1]
```diff
@@ -466,10 +466,8 @@ Class Procs:
 /obj/machinery/tesla_act(var/power)
 	..()
 	if(prob(85))
-		emp_act(2)
+		explosion(src.loc,1,2,4,flame_range = 2, adminlog = 0, smoke = 0)
```
Comment (ChangelingRain) [inline code/game/machinery/machinery.dm]: Like I said before, this should be the LEAST common effect, not the most. This makes the tesla absolutely rip the station in half, like what the hell.

**Comment by Grandsteel** [comment]
I cannot tell you my disappointment with old Tesla, when it was just a mass murdering machine IF it even moves around to people. Exploding consoles and other electronics manages to actually give it some radical damage. Also complaints about it's damage being too strong feel invalid, since you have to compare it to singularity, wherein Tesla is much less damage since explosions are only around machinery, which can be avoided. Also Tesla is still kinda mediocre to begin with since it tends to get stuck in certain areas, the only reason it absolutely fucked the station in the ass was because penguin set the Tesla to home in on people that prayed at a certain time.

**Comment by Shadowlight213** [comment]
@PKPenguin321 The explosions apply to the tesla zaps from throwing stuff at shocked grills.
Someone threw a toolbox and blew up the bridge.

**Comment by ShadowDeath6** [comment]
why were tesla grilles even added, they're almost an instakill even when they're low power

File: code/game/machinery/machinery.dm (hunk 1 above)
Comment (PKPenguin321) [inline code/game/machinery/machinery.dm]: >This makes the tesla absolutely rip the station in half

Rip the station in half?
You mean like the singularity?

Oh, good, working as intended.

Comment (ShadowDeath6) [inline code/game/machinery/machinery.dm]: didn't people not want this to be a copy paste of the singlo? I get boosting its devastation but making it explode stuff pretty much every time it zaps would be kind of lame

Comment (Iamgoofball) [inline code/game/machinery/machinery.dm]: yeah this is not supposed to just be a copypasta of singularity

**Comment by PKPenguin321** [comment]
@Iamgoofball I don't want to come off as rude, but the players in deadchat absolutely loved what I did to the tesla. It might go against your "design vision," but I heard time and time again that the tesla is now "much more fun to watch." You could ask almost anyone and they would agree that the current tesla we have right now is very, very boring, *especially* compared to the singularity.

This is just outright better and more popular/fun to the players, and there's no denying that.

File: code/game/machinery/machinery.dm (hunk 1 above)
Comment (yackemflam) [inline code/game/machinery/machinery.dm]: The tesla was supposed to be a lag free engine if I remember correctly

**Comment by Iamgoofball** [comment]
I guess. Just see if you can make the explosion visuals a little more electrified.

File: code/game/machinery/machinery.dm (hunk 1 above)
Comment (PKPenguin321) [inline code/game/machinery/machinery.dm]: This is lag free if you'd believe it

**Comment by PKPenguin321** [comment]
In this commit:
- Refactored tesla grilles. They no longer shoot a real tesla beam, just one that looks exactly like a tesla beam and will ONLY target carbon mobs (meaning no explosions).

**Comment by Iamgoofball** [comment]
:-1: no no no

they need to use proper tesla zap code, don't snowflake it

just disable the explosion for the tesla grille

**Comment by PKPenguin321** [comment]
I tried like 40 different methods and spent two days on that, this was the best I could do. If you have an actual code-based method for removing the explosion for tesla grille, please please let me hear it.

**Comment by Iamgoofball** [comment]
i'll look into it fam dont worry

**Comment by PKPenguin321** [comment]
In this commit:
- Made the gravity generator get properly destroyed by the tesla so that it doesn't just repeatedly explode over and over.
- Made the nuke do the same as the gravity generator. As a consequence, nukes can now be "disarmed" by the tesla (in the same way BoH bombing a nuke disarms it).

**Comment by PKPenguin321** [comment]
@Jordie0608 @Razharas @optimumtact @ChangelingRain @RemieRichards @tkdrg @Dorsisdwarf @KorPhaeron 
First person to merge this is confirmed cool

**Comment by RemieRichards** [comment]
I'm already cool though.

**Comment by PKPenguin321** [comment]
But are you confirmed cool

**Comment by RemieRichards** [comment]
yes.

**Comment by PKPenguin321** [comment]
shoot well could you still merge this please

**Comment by optimumtact** [comment]
Time to make space great again.

**Pull request merged**